src/WLR/Types/Seat.hsc

```diff
@@ -0,0 +1,425 @@
+    keyboard, Ptr WLR_keyboard,
+    focused_client, Ptr WLR_seat_client,
+    focused_surface, Ptr WLR_surface,
+    keyboard_destroy, WL_listener,
+    keyboard_keymap, WL_listener,
+    keyboard_repeat_info, WL_listener,
+    surface_destroy, WL_listener,
+    grab, Ptr WLR_seat_keyboard_grab,
+    default_grab, Ptr WLR_seat_keyboard_grab,
+    events focus_change, WL_signal,
+}}
+
+{{ struct
+    wlr/types/wlr_seat.h,
+    wlr_seat_touch_state,
+    seat, Ptr WLR_seat,
+    touch_points, WL_list,
+    grab_serial, Word32,
+    grab_id, Word32,
+    grab, Ptr WLR_seat_touch_grab,
+    default_grab, Ptr WLR_seat_touch_grab,
+}}
+
+{{ struct
+    wlr/types/wlr_seat.h,
+    wlr_seat,
+    global, Ptr WL_global,
+    display, Ptr WL_display,
+    clients, Ptr WL_list,
+    name, CString,
+    capabilities, Word32,
+    accumulated_capabilities, Word32,
+    last_event, TIMESPEC,
+    selection_source, Ptr WLR_data_source,
+    selection_serial, Word32,
+    selection_offers, WL_list,
+    primary_selection_source, Ptr WLR_primary_selection_source,
+    primary_selection_serial, Word32,
+    drag, Ptr WLR_drag,
+    drag_source, Ptr WLR_data_source,
+    drag_serial, Word32,
+    drag_offers, WL_list,
+    pointer_state, WLR_seat_pointer_state,
+    keyboard_state, WLR_seat_keyboard_state,
+    touch_state, WLR_seat_touch_state,
+    display_destroy, WL_listener,
+    selection_source_destroy, WL_listener,
+    primary_selection_source_destroy, WL_listener,
+    drag_source_destroy, WL_listener,
+    events pointer_grab_begin, WL_signal,
+    events pointer_grab_end, WL_signal,
+    events keyboard_grab_begin, WL_signal,
+    events keyboard_grab_end, WL_signal,
+    events touch_grab_begin, WL_signal,
+    events touch_grab_end, WL_signal,
+    events request_set_cursor, WL_signal,
+    events request_set_selection, WL_signal,
+    events set_selection, WL_signal,
+    events request_set_primary_selection, WL_signal,
+    events set_primary_selection, WL_signal,
+    events request_start_drag, WL_signal,
+    events start_drag, WL_signal,
+    events destroy, WL_signal,
+    data, Ptr ()
+}}
+
+{{ struct
+    wlr/types/seat.h,
+    wlr_touch_point,
+    touch_id, Int32,
+    surface, Ptr WLR_surface,
+    client, Ptr WLR_seat_client,
+    focus_surface, Ptr WLR_surface,
+    focus_client, Ptr WLR_seat_client,
+    sx, CDouble,
+    sy, CDouble,
+    surface_destroy, WL_listener,
+    focus_surface_destroy, WL_listener,
+    client_destroy, WL_listener,
+    events destroy, WL_signal,
+    link, WL_list
+}}
+
+-- TODO interface was a ConstPtr
+-- it looks there is a type for this starting in base 4.18.0.0
+-- https://hackage.haskell.org/package/base-4.19.0.0/docs/Foreign-C-ConstPtr.html
+
+{{ struct
+    wlr/types/wlr_seat.h,
+    wlr_seat_keyboard_grab,
+    interface, Ptr WLR_keyboard_grab_interface,
+    seat, Ptr WLR_seat,
+    data, Ptr ()
+}}
+
+-- TODO enter's Word32 final parameter is a const array
+-- how do I do 'const' in HSC?
```

Comment:
As far as I’m aware, there should be no need to worry about constness on the Haskell side. (`ConstPtr` exists, but as you note that requires a very recent version of `base`.)